File: Web Survey/resources/js/recommender/distance_metrics/PearsonCorrelation.js
import AbstractDistanceMetric from "./AbstractDistanceMetric";

export default class PearsonCorrelation extends AbstractDistanceMetric {

    static getMetricName() {
        return "Pearson Correlation";
    }

    calculateMean(track, attributes) {
        const self = this;

        let mean = 0;
        this.attributeKeys.forEach(attribute => {
            if(attribute === "loudness") {
                mean += self.fixLoudness(attributes[track][attribute]);
            } else {
                mean += attributes[track][attribute];
            }
        });
        mean /= this.attributeKeys.length;

        return mean;
    }

    calculateDistance(track, otherTrack, attributes) {
        const self = this;

        let distanceA = 0;
        let distanceB = 0;
        let distanceAB = 0;

        let meanTrack = this.calculateMean(track, attributes);
        let meanOther = this.calculateMean(otherTrack, attributes);

        this.attributeKeys.forEach(attribute => {

            let attributeOwn;
            let attributeOther;

            if(attribute === "loudness") {
                attributeOwn = self.fixLoudness(attributes[track][attribute]);
                attributeOther = self.fixLoudness(attributes[otherTrack][attribute])
            } else {
                attributeOwn = attributes[track][attribute];
                attributeOther = attributes[otherTrack][attribute];
            }

            distanceA += (attributeOther - meanOther) ** 2;
            distanceB += (attributeOwn - meanTrack) ** 2;
            distanceAB += (attributeOther - meanOther) * (attributeOwn - meanTrack);
        });
        let distance;

        try {
            distance = distanceAB / Math.sqrt(distanceA * distanceB)
        } catch(err){
            console.log("Zero division");
            return NaN;
        }
        return distance;
    }

}

File: Experiment Processing/recommender/distance_metrics/pearson_correlation.py
import math
import numpy as np

from recommender.distance_metrics.abstract_distance_metric import AbstractDistanceMetric


class PearsonCorrelation(AbstractDistanceMetric):
    """
    Pearson correlation, see https://en.wikipedia.org/wiki/Pearson_correlation_coefficient for more information.
    Roughly mirrors the correlation distance.
    """

    def get_name(self):
        return "Pearson Correlation"

    def calculate_mean(self, track_features):

        mean = 0

        for attribute in self.attribute_keys:
            mean += track_features[attribute]

        return mean / len(self.attribute_keys)

    def calculate_distance(self, track_a, track_b):
        distance_a = 0
        distance_b = 0
        distance_a_b = 0

        track_a_features = track_a.get_metadata()
        track_b_features = track_b.get_metadata()

        mean_a = self.calculate_mean(track_a_features)
        mean_b = self.calculate_mean(track_b_features)

        for attribute in self.attribute_keys:
            attribute_a = track_a_features[attribute]
            attribute_b = track_b_features[attribute]

            distance_a += (attribute_a - mean_a) ** 2
            distance_b += (attribute_b - mean_b) ** 2
            distance_a_b += (attribute_b - mean_b) * (mean_a - mean_a)

        try:
            return distance_a_b / (math.sqrt(distance_a * distance_b))
        except ZeroDivisionError:
            return np.NaN
